fix python rpc handler exit crash

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27251 fix python rpc handler exit crash**

[test all] 

Explicitly clean up py::objects to avoid segment faults when py::objects with CPython are cleaned up later at program exit.

See similar issues reported https://github.com/pybind/pybind11/issues/1598
and https://github.com/pybind/pybind11/issues/1493.

Our local tests also caught this segment faults if py::objects are cleaned
up at program exit. The explaination is: CPython cleans up most critical
utitlies before cleaning up PythonRpcHandler singleton, so when
PythonRpcHandler signleton cleans up py::objects and call dec_ref(), it
will crash.

The solution is to clean up py::objects earlier when Rpc agent join().
Be note that py::objects can not be cleaned up when Rpc agent is destroyed
as well, as Rpc agent is global variable and it will have same issue as
PythonRpcHandler.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

zhaojuanmao committed:
fix python rpc handler exit crash in python 3.5 spwan tests

we found python rpc handler exit crash in python 3.5 spwan tests, backtrace showed pythonRpcHandler destructor dec_ref() deallocated python objects, that means python objects in pythonRpcHandler have been cleaned up before destructing pythonRpcHandler while python program exits in python 3.5.

This diff is to explicitly clean up python objects of pythonRpcHandler in RpcAgent.join(). so that pythonRpcHandler destructor will not have chances to call dec_ref(). we did not explicitly clean up python objects in RpcAngent destructor, it is also because python objects will be cleaned up before destructing RpcAgent when python program exit in python 3.5.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -157,6 +157,12 @@ void ProcessGroupAgent::join() {
+  // the python objects in PythonRpcHandler could be cleaned up before
+  // destructing ProcessGroupAgent, then ProcessGroupAgent destructor will
+  // dec_ref deallocated python objects and crash.
+  PythonRpcHandler::getInstance().cleanUp();
```
What if we just set `runUDFFunction_` and `loadResultFunction_` to None in the destructor of PythonRPCHandler? Does that work?

zhaojuanmao replied:
if we only set them to None, not free the memory, then we are assuming python objects will be cleaned up before destructing pythonRpcHandler, but this assumption is only true for python 3.5. For python 3.6, it is not the case, we may be at risk of leaking memory in python 3.6 if set them as None only in destructor of PythonRPCHandler

pritamdamania87 replied:
IIUC, PythonRpcHandler is a singleton so it would be destructed only on program exit. As a result, we won't leak any memory since the OS will just clean up all the memory once the process dies. Although, I'm not sure if ASAN would fail due to this. Can you give it a shot?

zhaojuanmao replied:
@pritamdamania87 tests passed in both py3.5 and py3.6, fork asan did not complain as well, not sure whether spawn asan will complain or not because we disabled spawn asan. 

I will update diff to see how others think about it as well.

zhaojuanmao committed:
Update on "fix python rpc handler exit crash in python 3.5 spwan tests"


we found python rpc handler exit crash in python 3.5 spwan tests, backtrace showed pythonRpcHandler destructor dec_ref() deallocated python objects, that means python objects in pythonRpcHandler have been cleaned up before destructing pythonRpcHandler while python program exits in python 3.5.

This diff is to explicitly clean up python objects of pythonRpcHandler in RpcAgent.join(). so that pythonRpcHandler destructor will not have chances to call dec_ref(). we did not explicitly clean up python objects in RpcAngent destructor, it is also because python objects will be cleaned up before destructing RpcAgent when python program exit in python 3.5.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -179,6 +179,12 @@ void ProcessGroupAgent::join() {
+  // ProcessGroupAgent. Otherwise, in Python 3.5, when python program exits,
+  // the python objects in PythonRpcHandler could be cleaned up before
+  // destructing ProcessGroupAgent, then ProcessGroupAgent destructor will
+  // dec_ref deallocated python objects and crash.
```
Question: do we know why `inc_ref` was not called properly in the first place? If `ProcessGroupAgent` is holding a reference as well, should the reference counting be consistent with that?

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -179,6 +179,12 @@ void ProcessGroupAgent::join() {
       SendWork(workerIds_[dst], Message({}, {}, MessageType::SHUTDOWN)));
   threadPool_.waitWorkComplete();
   listenerThread_.join();
+  // explicitly clean up python objects in PythonRpcHandler before destructing
```
If we expect all `RpcAgent` implementations would use the same `PythonRpcHandler`, shall we add this to `RpcAgent` instead of an implementation of `RpcAgent`.

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -12,6 +12,18 @@ PythonRpcHandler::PythonRpcHandler() {
+void PythonRpcHandler::cleanUp() {
+  AutoGIL ag;
+  if (!runUDFFunction_.is_none()) {
+    runUDFFunction_.dec_ref();
```
Could you please remind me where is the corresponding `inc_ref` for this? Is it when we load the function in `module.attr`? I am a little confused of why we need to do `dec_ref` on our own but not `inc_ref`?

zhaojuanmao commented:
@mrshenli the theory now is: inc_ref works as expected, but python objects are deallocated before destructing pythonRpcHandler when program exits, so when pythonRpcHandler destructor tries to call dec_ref to clean up python objects, it is trying to clean up null objects and crashed. 

And this only happens to python 3.5, I do not know why python objects are cleaned up before destructing pythonRpcHandler in python 3.5 only.  This is not the case for python 3.6 though. 

Does it make sense?

zhaojuanmao committed:
Update on "fix python rpc handler exit crash in python 3.5 spwan tests"


we found python rpc handler exit crash in python 3.5 spwan tests, backtrace showed pythonRpcHandler destructor dec_ref() deallocated python objects, that means python objects in pythonRpcHandler have been cleaned up before destructing pythonRpcHandler while python program exits in python 3.5.

This diff is to explicitly clean up python objects of pythonRpcHandler in RpcAgent.join(). so that pythonRpcHandler destructor will not have chances to call dec_ref(). we did not explicitly clean up python objects in RpcAngent destructor, it is also because python objects will be cleaned up before destructing RpcAgent when python program exit in python 3.5.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]

mrshenli commented:
> inc_ref works as expected, but python objects are deallocated before destructing pythonRpcHandler when program exits, so when pythonRpcHandler destructor tries to call dec_ref to clean up python objects, it is trying to clean up null objects and crashed.

I am not aware of the root cause for this, but this Python deallocation behavior seems weird to me. If `pythonRpcHandler` is still holding a reference to the Python object, wouldn't that Python object has a non-zero ref-count? If so, I am confused why they are deallocated before `pythonRpcHandler`. I would expect Python walk through the reference graph to destruct things in order. Or maybe Python just don't respect the topological order just in case there are circular dependencies.

zhaojuanmao commented:
@mrshenli right, that is also weird to me as well. But be noted, Python 3.6 does things as we expected. but python 3.5 does not

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -13,6 +13,24 @@ PythonRpcHandler::PythonRpcHandler() {
+// PythonRpcHandler singleton when program exits, all memories will be cleaned
+// up by OS.
+PythonRpcHandler::~PythonRpcHandler() {
+  if (!runUDFFunction_.is_none()) {
```
nit: Do we need these if statements? Can't we just set this to `py::none()`?

zhaojuanmao committed:
Update on "fix python rpc handler exit crash in python 3.5 spwan tests"


we found python rpc handler exit crash in python 3.5 spwan tests, backtrace showed pythonRpcHandler destructor dec_ref() deallocated python objects, that means python objects in pythonRpcHandler have been cleaned up before destructing pythonRpcHandler while python program exits in python 3.5.

This diff is to explicitly clean up python objects of pythonRpcHandler in RpcAgent.join(). so that pythonRpcHandler destructor will not have chances to call dec_ref(). we did not explicitly clean up python objects in RpcAngent destructor, it is also because python objects will be cleaned up before destructing RpcAgent when python program exit in python 3.5.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]

mrshenli reviewed (approved):
There might be conflicts with #27284.

zhaojuanmao committed:
Update on "fix python rpc handler exit crash in python 3.5 spwan tests"


we found python rpc handler exit crash in python 3.5 spwan tests, backtrace showed pythonRpcHandler destructor dec_ref() deallocated python objects, that means python objects in pythonRpcHandler have been cleaned up before destructing pythonRpcHandler while python program exits in python 3.5.

This diff is to explicitly clean up python objects of pythonRpcHandler in RpcAgent.join(). so that pythonRpcHandler destructor will not have chances to call dec_ref(). we did not explicitly clean up python objects in RpcAngent destructor, it is also because python objects will be cleaned up before destructing RpcAgent when python program exit in python 3.5.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -26,6 +26,21 @@ PythonRpcHandler::PythonRpcHandler() {
+  pyLoadReturnValue_ = py::none();
+  pySerialize_ = py::none();
+}
+
```
Won't assigning them to be `py::none()` call `dec_ref` on the existing value?

zhaojuanmao replied:
if I understand correctly, assigning py::none() will not trigger dec_ref() call.  in default, when py::object is destroyed, dec_ref() will be called first and then assign it as py::none()?

pietern reviewed (changes requested):
The two checks were carried over from the previous diff but are no longer needed.

The explanation of why this works on Python 3.5 is counter intuitive to me.

If Python 3.5 is proving to be such a pain, can't we restrict RPC to 3.6?

zhaojuanmao commented on test/run_test.py:
```diff
@@ -28,6 +29,7 @@
     'cuda_primary_ctx',
     'dataloader',
     'dist_autograd_fork',
+    'dist_autograd_spawn',
```
right, I was waiting to rebase on your diff

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -26,6 +26,21 @@ PythonRpcHandler::PythonRpcHandler() {
+// PythonRpcHandler singleton when program exits, all memories will be cleaned
+// up by OS.
+PythonRpcHandler::~PythonRpcHandler() {
+  pyRunFunction_ = py::none();
```
Do we need GIL here?

zhaojuanmao replied:
no, I tired to use GIL, it will throw error

mrshenli replied:
what error did you see? It sounds suspicious that adding synchronization triggers errors.

zhaojuanmao replied:
my understanding is: at this point, python interpreter may be gone?

zhaojuanmao replied:
some python system objects and states are not there anymore at this point

mrshenli replied:
I see

zhaojuanmao committed:
Update on "fix python rpc handler exit crash in python 3.5 spwan tests"


[test all] we found python rpc handler exit crash in python 3.5 spwan tests, backtrace showed pythonRpcHandler destructor dec_ref() deallocated python objects, that means python objects in pythonRpcHandler have been cleaned up before destructing pythonRpcHandler while python program exits in python 3.5.

This diff is to explicitly clean up python objects of pythonRpcHandler in RpcAgent.join(). so that pythonRpcHandler destructor will not have chances to call dec_ref(). we did not explicitly clean up python objects in RpcAngent destructor, it is also because python objects will be cleaned up before destructing RpcAgent when python program exit in python 3.5.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "fix python rpc handler exit crash "


[test all] 

Explicitly clean up py::objects to avoid segment faults when py::objects with CPython are cleaned up later at program exit.

See similar issues reported https://github.com/pybind/pybind11/issues/1598
and https://github.com/pybind/pybind11/issues/1493.

Our local tests also caught this segment faults if py::objects are cleaned
up at program exit. The explaination is: CPython cleans up most critical
utitlies before cleaning up PythonRpcHandler singleton, so when
PythonRpcHandler signleton cleans up py::objects and call dec_ref(), it
will crash.

The solution is to clean up py::objects earlier when Rpc agent join().
Be note that py::objects can not be cleaned up when Rpc agent is destroyed
as well, as Rpc agent is global variable and it will have same issue as
PythonRpcHandler.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -16,6 +16,19 @@ py::object getFunction(const py::object& module, const char* name) {
 
+// if cachedFn is not none, assign it to local function; otherwise, import
+// function from the scratch and assign it to local function;
+py::object getLocalFunction(const py::object& cachedFn, const char* name) {
```
why do we need this given that the three py functions are already loaded in the constructor?

zhaojuanmao replied:
because the three py functions will be cleaned up in Rpc agent join(), but they could be called after Rpc agent join() in rare cases. 

for most cases, we do not need to load them again, just copy cachedFn to localFn, this should be cheap as it is just copying pointer. 

Does it make sense?

mrshenli replied:
> but they could be called after Rpc agent join() in rare cases.

when could this happen? And if we load them again, do we need to cleanup them again?

zhaojuanmao replied:
e.g.  loadPythonUDFResult() is called in future.wait(), which could be called after Rpc agent join(). 

it does not need to clean up them again, as it is loaded to a local variable intentionally for this rare case

mrshenli replied:
I see. Let's get this PR in to fix master, and let's discuss whether we need to support the above case. I feel that all `future.wait()` in user code should be resolved before calling `join`, and if we detect termination in `join()`, `RequestCallbackImpl` won't hit this case either (?).

zhaojuanmao replied:
@mrshenli yes, RequestCallbackImpl will not hit this case. but I still add it for runPythonUDF() function for safety reason, because in theory functions in PythonRpcHandler could be called anytime.

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -16,6 +16,19 @@ py::object getFunction(const py::object& module, const char* name) {
+    localFn = cachedFn;
+  }
+  return localFn;
+}
```
How about:

```c++
  if (cachedFn.is_none()) {
    return getFunction(py::module::import("torch.distributed.rpc.internal"), name);
  } else {
    return cachedFn;
  }
```

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -39,6 +39,21 @@ class PYBIND11_EXPORT PythonRpcHandler {
+  // and https://github.com/pybind/pybind11/issues/1493
+  // Our local tests also caught this segment faults if py::objects are cleaned
+  // up at program exit. The explaination is: CPython cleans up most critical
+  // utitlies before cleaning up PythonRpcHandler singleton, so when
```
utitlies -> utilities ?

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -39,6 +39,21 @@ class PYBIND11_EXPORT PythonRpcHandler {
+  // The solution is to clean up py::objects earlier when Rpc agent join().
+  // Be note that py::objects can not be cleaned up when Rpc agent is destroyed
+  // as well, as Rpc agent is global variable and it will have same issue as
+  // PythonRpcHandler.
```
Just make sure I understand the above statement. Are you saying that we cannot put `cleanup` in `~ProcessGroupAgent` due to the same issue?

zhaojuanmao replied:
yes, I tried it before, it has same issue, because at that point, Cpython may clean up some critical utilities and not able to clean up the py::objects correctly sometimes

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -16,6 +16,19 @@ py::object getFunction(const py::object& module, const char* name) {
+    localFn =
+        getFunction(py::module::import("torch.distributed.rpc.internal"), name);
+  } else {
+    localFn = cachedFn;
```
(Does not need to be fixed in this PR) This will always make a copy of the py::object even if the cachedFn is available. Do you know how expensive it is to make a copy of py::object? It is just similar as creating a shared_ptr? Will it work if we return a const reference here to extend the lifetime of a temporary var?

zhaojuanmao replied:
if I understand correctly, py::object is a pointer, so should be cheap. I would rather not to return a const reference to extend lifetime, because I do not know when the const will be destroyed, may hit the same issue as now

zhaojuanmao committed:
Update on "fix python rpc handler exit crash "


[test all] 

Explicitly clean up py::objects to avoid segment faults when py::objects with CPython are cleaned up later at program exit.

See similar issues reported https://github.com/pybind/pybind11/issues/1598
and https://github.com/pybind/pybind11/issues/1493.

Our local tests also caught this segment faults if py::objects are cleaned
up at program exit. The explaination is: CPython cleans up most critical
utitlies before cleaning up PythonRpcHandler singleton, so when
PythonRpcHandler signleton cleans up py::objects and call dec_ref(), it
will crash.

The solution is to clean up py::objects earlier when Rpc agent join().
Be note that py::objects can not be cleaned up when Rpc agent is destroyed
as well, as Rpc agent is global variable and it will have same issue as
PythonRpcHandler.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]

mrshenli reviewed (approved):
The latest version looks good!

zhaojuanmao commented:
failures are not related

zhaojuanmao committed:
Update on "fix python rpc handler exit crash "


[test all] 

Explicitly clean up py::objects to avoid segment faults when py::objects with CPython are cleaned up later at program exit.

See similar issues reported https://github.com/pybind/pybind11/issues/1598
and https://github.com/pybind/pybind11/issues/1493.

Our local tests also caught this segment faults if py::objects are cleaned
up at program exit. The explaination is: CPython cleans up most critical
utitlies before cleaning up PythonRpcHandler singleton, so when
PythonRpcHandler signleton cleans up py::objects and call dec_ref(), it
will crash.

The solution is to clean up py::objects earlier when Rpc agent join().
Be note that py::objects can not be cleaned up when Rpc agent is destroyed
as well, as Rpc agent is global variable and it will have same issue as
PythonRpcHandler.

close #27182

Differential Revision: [D17727362](https://our.internmc.facebook.com/intern/diff/D17727362/)

[ghstack-poisoned]
